src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+        :return: 2-tuple of numpy arrays (Estimated average of defocus and an internal variable)
+        """
+
+        center = N // 2
+        [X, Y] = np.meshgrid(
+            np.arange(0 - center, N - center, dtype=self.dtype) / N,
+            np.arange(0 - center, N - center, dtype=self.dtype) / N,
+        )
+
+        rb = np.sqrt(np.square(X) + np.square(Y))
+        rb = rb[center, center:]
+        r_ctf = rb * (10 / pixel_size)
+
+        signal = thon_rings
+        signal = np.where(signal < 0, 0, signal)
+        signal = np.sqrt(signal)
+        signal = signal[: 3 * signal.shape[0] // 4]
+
+        r_ctf_sq = np.square(r_ctf)
+        c = np.zeros((9500, 13), dtype=self.dtype)
+
+        for f in range(500, 10000):
+            ctf_im = np.abs(
+                np.sin(
+                    np.pi * lmbd * f * r_ctf_sq
+                    - 0.5 * np.pi * (lmbd ** 3) * cs * (10 ** 6) * np.square(r_ctf_sq)
+                    + w
+                )
+            )
+            ctf_im = ctf_im[: signal.shape[0]]
+
+            ctf_im = np.reshape(ctf_im, (ctf_im.shape[0], 1))
+            ctf_im = np.tile(ctf_im, (1, 13))
+
+            for m in range(0, 13):
+                signal[:, m] = signal[:, m] - np.mean(signal[m + 1 :, m], axis=0)
+                ctf_im[:, m] = ctf_im[:, m] - np.mean(ctf_im[m + 1 :, m], axis=0)
+                ctf_im[: m + 1, m] = np.zeros((m + 1))
+                signal[: m + 1, m] = np.zeros((m + 1))
+
+            Sx = np.sqrt(np.sum(np.square(ctf_im), axis=0))
+            Sy = np.sqrt(np.sum(np.square(signal), axis=0))
+            c[f - 500, :] = np.divide(
+                np.sum(np.multiply(ctf_im, signal), axis=0), (np.multiply(Sx, Sy))
+            )
+
+        max_c = np.argmax(c)
+        arr_max = np.unravel_index(max_c, c.shape)
+        avg_defocus = arr_max[0] + 500
+        max_col = arr_max[1]
+
+        return avg_defocus, max_col
+
+    def ctf_background_subtract_2d(self, signal, background_p1, max_col):
+        """
+        Subtract background from estimated power spectrum
+
+        :param signal: Estimated power spectrum
+        :param background_p1: 1-D background estimation
+        :param max_col: Internal variable, returned as the second parameter from ctf_opt1d.
+        :return: 2-tuple of numpy arrays (Estimated PSD without noise and estimated noise).
+        """
+
+        # RCOPT
+        signal = signal.asnumpy().T
+        # background_p1 is still np array in old ordering for now.
+
+        N = signal.shape[0]
+        center = N // 2
+        [X, Y] = np.meshgrid(
+            np.arange(0 - center, N - center, dtype=self.dtype),
+            np.arange(0 - center, N - center, dtype=self.dtype),
+        )
+        radii = np.sqrt(X ** 2 + Y ** 2)
+
+        background = np.zeros(signal.shape, dtype=self.dtype)
+        background_p1 = background_p1[:, max_col]
+        # background = thon_rings.copy()
```

Comment:
Noting I'll rm this stray comment along with other requested changes that are required from the review.